src/persist/src/indexed/runtime.rs

```diff
@@ -323,6 +350,63 @@ impl<K: Clone, V: Clone> StreamWriteHandle<K, V> {
     }
 }
 
+/// A handle for atomically writing to multiple streams.
+pub struct AtomicWriteHandle<K, V> {
+    ids: HashSet<Id>,
+    runtime: RuntimeClient<K, V>,
+}
+
+impl<K: Clone, V: Clone> AtomicWriteHandle<K, V> {
+    /// Returns a new [AtomicWriteHandle] for the given streams.
+    pub fn new(handles: &[&StreamWriteHandle<K, V>]) -> Result<Self, Error> {
+        let mut stream_ids = HashSet::new();
+        let runtime = if let Some(handle) = handles.first() {
+            handle.runtime.clone()
+        } else {
+            return Err(Error::from("AtomicWriteHandle received no streams"));
+        };
+        for handle in handles.iter() {
+            if handle.runtime.id != runtime.id {
+                return Err(Error::from(format!(
+                    "AtomicWriteHandle got handles from two runtimes: {:?} and {:?}",
+                    runtime.id, handle.runtime.id
+                )));
+            }
+            // It's odd if there are duplicates but the semantics of what that
+            // means are straightforward, so for now we support it.
+            stream_ids.insert(handle.id);
+        }
+        Ok(AtomicWriteHandle {
+            ids: stream_ids,
+            runtime,
+        })
+    }
+
+    /// Atomically writes the given updates to the paired streams.
+    ///
+    /// Either all of the writes will be made durable for replay or none of them
+    /// will.
+    //
+    // TODO: This could take &StreamWriteHandle instead of Id to avoid surfacing
```

Comment:
could also take a string stream name and then do the translation downstream? idk if thats worth it

Reply:
StreamWriteHandle has already done the string->id lookup, so that's probably better than the string name. Either way, I'm okay with the abstraction breakage for now